sklearn/cluster/hierarchical.py

```diff
@@ -696,7 +717,15 @@ def fit(self, X, y=None):
                              " instance, got 'memory={!r}' instead.".format(
                                  type(memory)))
 
-        if self.n_clusters <= 0:
+        if (self.distance_threshold and self.n_clusters) or \
+                (not self.n_clusters and not self.distance_threshold):
+            raise ValueError("Either n_clusters (>0) or distance_threshold "
+                             "needs to be set, got n_clusters={} and "
+                             "distance_threshold={} instead. "
+                             "Please set n_clusters=None to continue.".format(
```

Comment:
Instead of setting the `n_clusters`, we can make it a private attribute at __init__ like

```py
self._n_clusters = n_clusters
```

and have a property to define it's actual value like

```py
@property
def n_clusters(self):
    if self._n_clusters is None and self.distance_threshold is None:
        return 2
    elif self.distance_threshold is None ^ self._n_clusters is None:
        return self._n_clusters
    else:
        raise AttributeError("Both n_clusters and distance_threshold parameters"
                             "were set during initialization.")
```